pandas/tests/frame/indexing/test_at.py

```diff
@@ -0,0 +1,10 @@
+from datetime import datetime, timezone
+
+import pandas as pd
+
+
+def test_at_timezone():
+    # https://github.com/pandas-dev/pandas/issues/33544
+    df = pd.DataFrame({"foo": [datetime(2000, 1, 1)]})
+    df.at[0, "foo"] = datetime(2000, 1, 2, tzinfo=timezone.utc)
+    assert df.at[0, "foo"].tzinfo == timezone.utc
```

Comment:
Could you instead create an `expected` Dataframe and use `tm.assert_frame_equal`?

```
    result = pd.DataFrame({"foo": [datetime(2000, 1, 1)]})
    result.at[0, "foo"] = datetime(2000, 1, 2, tzinfo=timezone.utc)
    expected = pd.DataFrame(....)
    tm.assert_frame_equal(result, expected)
```

Reply:
Hi @mroeschke ,

I have implemented as you suggested, but actually (as you can see in the checks below) this makes the test fail and raises an interesting question.

When creating the dataframe with:
```python
result = pd.DataFrame({"foo": [datetime(2000, 1, 1)]})
```
It transforms the value `result.at[0, "foo"]` into a _Timestamp_ object:`Out[4]: Timestamp('2000-01-01 00:00:00')`and running `result.info()` will show that the column Dtype is `datetime64[ns]`.

If now I apply the following
```python
result.at[0, "foo"] = datetime(2000, 1, 2, tzinfo=timezone.utc)
```
the `result.at[0, "foo"]` is transformed in to a _datetime_ object
`Out[6]: datetime.datetime(2000, 1, 2, 0, 0, tzinfo=datetime.timezone.utc)` and also the column Dtype is changed to `object`.

I am not sure if this is the expected behaviour or it is a bug.

If this is the expected behaviour I should change the test, otherwise `.at` needs to be fixed.

Reply:
I _think_ this is the expected behavior as opposed to the `loc` variant that would coerce to datetime type.

You can therefore change the expected construction to

`expected = pd.DataFrame(np.array([datetime(2000, 1, 2, tzinfo=timezone.utc)], dtype=object, columns['foo'])`

Reply:
> I _think_ this is the expected behavior

this seems inconsisent with other types that cannot hold the scalar type.

```
>>> import numpy as np
>>>
>>> import pandas as pd
>>>
>>> pd.__version__
'1.1.0.dev0+1290.g7e99bb433'
>>>
>>> df = pd.DataFrame({"a": [1, 2, 3]})
>>> df
   a
0  1
1  2
2  3
>>>
>>> df.at[0, "a"] = np.nan
Traceback (most recent call last):
...
ValueError: Cannot assign nan to integer series
>>>
```